app/frontend/actions.ts

```diff
@@ -637,31 +632,42 @@ export default ({setState, getState}: {setState: SetStateFn; getState: GetStateF
     }
     const coins = state.sendAmount.coins
     const donationAmount = state.donationAmount.coins
+    // TODO: sendAddress should have a validated field of type _Address
     const address = state.sendAddress.fieldValue as _Address
-    const plan = await prepareTxPlan({address, coins, donationAmount, txType: TxType.SEND_ADA})
-    const newState = getState() // if the values changed meanwhile
-    if (
-      newState.sendAmount.fieldValue !== state.sendAmount.fieldValue ||
-      newState.sendAddress.fieldValue !== state.sendAddress.fieldValue ||
-      newState.donationAmount.fieldValue !== state.donationAmount.fieldValue
-    ) {
-      return
-    }
-    const validationError = txPlanValidator(
+    const txPlanResult = await prepareTxPlan({
+      address,
       coins,
-      getSourceAccountInfo(state).balance, // TODO: get new balance
-      plan,
-      donationAmount
-    )
-    setErrorState('sendAmountValidationError', validationError)
-    if (!validationError) {
-      setTransactionSummary('send', plan, coins, donationAmount)
+      donationAmount,
+      txType: TxType.SEND_ADA,
+    })
+    const balance = getSourceAccountInfo(state).balance as Lovelace
+
+    if (txPlanResult.type === TxPlanResultType.SUCCESS) {
+      const newState = getState() // if the values changed meanwhile
```

Comment:
I'm not that confident this check actually works - if I'm not mistaken, our state is currently mutable so the "old" state's nested keys will get modified regardless

Reply:
ok, no, I checked and the state storage implementation copies the top-level key-value pairs into a new object each time they are modified: https://github.com/vacuumlabs/adalite/blob/eebae33b7bbb339d96d02c1c320f9547778134bf/app/frontend/libs/unistore/index.js#L34

Reply:
:D I don't think it works either.. I ll test it and remove it. :+1: